### ISSUE TITLE
Add system constraints

### ISSUE THREAD

**Issue by mserajnik** [mserajnik/hydrusrv#16]
# System constraints

## Outline

hydrus client-style _system constraints_ should be added to hydrusrv. The goal is not to replicate everything available, but to provide a good set that further increases the users ability to filter their files.

As a first step, the following constraints should be added (I've borrowed their names from hydrus client, where possible):

+ `system:id`:  `=`, `!=`, `~=`, `>`, `<`, `><` (combines `>` and `<`, see [here](https://github.com/mserajnik/hydrusrv/issues/16#issuecomment-449524927) for reasoning)
+ `system:number of tags`: `=`, `!=`, `~=`, `>`, `<`, `><` (field needs to be added – see [here](https://github.com/mserajnik/hydrusrv-sync/issues/2))
+ `system:size`: `=`, `!=`, `~=`, `>`, `<`, `><`
+ `system:width`:  `=`, `!=`, `~=`, `>`, `<`, `><`
+ `system:height`:  `=`, `!=`, `~=`, `>`, `<`, `><`
+ `system:hash`: `=`, `!=`
+ `system:mime`: `=`, `!=`

For each of these fields, the user will need to provide a value to compare to by himself (no comparisons like `width > height` for now). For `system:hash` and `system:mime`,  those values will be strings (a SHA-256 for `hash` and the mime type in the common form, e.g. `image/png`, for `mime`) while all the other fields will takes integers.

`~=` compares _approximately_, e.g., `width~=1000` might return files that are `>800` and `<1200`. ~~If it's not too complicated (= slow), this should probably mirror the behavior in hydrus client (haven't looked at it yet). If it is, the range might just be a fixed percentage (e.g., 20%).~~ __Edit:__ See [here](https://github.com/mserajnik/hydrusrv/issues/16#issuecomment-446717289) for the proper solution.

The user can provide multiple constraints at once, each one limiting the set further (together with tags and exclude tags); no `OR` comparisons (for now). The one exception is if multiple constraints for the same field are provided, in which case all the constraints for that specific field are compared in an `OR` fashion.

## API

The API query parameters for this could looke like the following:

```constraints[]=<field><comparison operator><value>```

E.g.,

```constraints[]=width=1000```

or

```constraints[]=height>500```

or

```constraints[]=mime=image/png&constraints[]=width>2000```

In reality, many of these character probably need to URL-encoded (depending on the client), but I left that out to make it easier to read here.

## Frontend (hydrusrvue)

~~In hydrusrvue, this could be handled in two ways:~~

~~1. Use the existing tag input and add an option (button?) to set it to _system constraint mode_ (where the input will be autocompleted for constraints instead of tags). The user could then enter `width>1000` and add that constraint just like with tags. Also like tags, it could be displayed as such below the search form (but with a distinct color, icon and sorted either to the front or the back of the listed "tags")~~
~~2. Add a new mask, likely a modal or quickview, where the user can add/remove constraints. This could be split across multiple inputs (e.g., a dropdown to select a field, a dropdown to select the operator and a text input to enter the value)~~

~~Imo, 1.) can work really well, but needs to be carefully done not to confuse people and not make the tag input worse (it should always default to tag input and there should be no way to switch to constraint mode by accident). 2.) is probably easier to understand at first, but more cumbersome and likely slower for "power users" (the user needs to open a separate view).~~

~~More considerations about the frontend once the server-side is done.~~

See [here](https://github.com/mserajnik/hydrusrv/issues/16#issuecomment-446733754) for the solution I'll go with.

## Timeframe

Development will likely not start before January due to busy holidays. The goal is to finish this (including the necessary frontend additions) by the end of January.

**Comment by FredericaBernkastel** [comment]
> ≈ compares approximately
>  If it is, the range might just be a fixed percentage (e.g., 20%).

If i remember correctly, the only proper way is to use probability density function of [normal distribution](https://en.wikipedia.org/wiki/Normal_distribution).

![clipboard-110](https://user-images.githubusercontent.com/16236235/49892842-2aa64000-fe74-11e8-9731-1700ab2bea11.png)

I did some math, and it works perfectly, as i expected.
Some pseudocode:
```
μ = 0;
σ = 1;
ε = 0.618;
x = 10;

distribution = NormalDistribution[μ, σ];
theta = x / PDF[distribution, 0];

deviance = (PDF[distribution, 0] - PDF[distribution, ε]) * theta;

lower = Round[x - deviance]
upper = Round[x + deviance]

Plot[PDF[distribution, x] * theta, {x, -3.14, 3.14}]
```
Where **`x`** is the number we're comparing to approximately.
By modifying **`ε`** you can get wider or smaller [Epsilon-neighborhood](https://en.wikipedia.org/wiki/Neighbourhood_%28mathematics%29).

**Examples:**

**`x = 4 -> [3, 5]`**
**`x = 6 -> [5, 7]`**
**`x = 10 -> [8, 12]`**
**`x = 100 -> [83, 117]`**
**`x = 1000 -> [826, 1174]`**

![epsilon-neighborhood](https://user-images.githubusercontent.com/16236235/49893686-3bf04c00-fe76-11e8-804d-15c97cb1fa67.png)
Where **`ε = 0.618`**

<br>

**`x = 4 -> [2, 6]`**
**`x = 6 -> [8, 4]`**
**`x = 10 -> [6, 14]`**
**`x = 100 -> [61, 139]`**
**`x = 1000 -> [607, 1393]`**

![epsilon-neighborhood](https://user-images.githubusercontent.com/16236235/49894254-b40b4180-fe77-11e8-99c3-defa93ee9cb0.png)
Where **`ε = 1`**

**Comment by mserajnik** [comment]
Interesting, thanks!

**Comment by FredericaBernkastel** [comment]
> a SHA-256 for hash

There should be both MD5 and SHA-256 support, since MD5 is a de-facto standard for all booru websites (deduplication, source tracking).

**Comment by mserajnik** [comment]
> > a SHA-256 for hash
> 
> There should be both MD5 and SHA-256 support, since MD5 is a de-facto standard for all booru websites (deduplication, source tracking).

But in that case we'd need to generate the MD5 ourselves since hydrus server only stores the SHA-256 iirc. Could increase the sync time _a lot_.

**Comment by FredericaBernkastel** [comment]
> Frontend (hydrusrvue)

I'd prefer using existing tag input with dropdown:

User starts typing **`"sys"`** -> show dropdown with **`"system"`** namespace
User starts typing constraint type **`"system:n"`** -> show dropdown with **`"system:number of tags"`** constraint.
User done typing constraint -> show dropdown with possible operators
User done typing operator -> show dropdown with possible constraint arguments (in case of mime).
User done typing constraint argument -> finish input and send the query.

Alot of benefits this way: you can simply copy-paste constraints, no special popups.

Also, i'd want to see added constraints as regular tags (like excluded tag, but with this [cute icon](https://fontawesome.com/icons/function?style=solid))

To help users, add help page with all possible queries (see [danbooru cheatsheet](https://danbooru.donmai.us/wiki_pages/43049))

**Comment by FredericaBernkastel** [comment]
> But in that case we'd need to generate the MD5 ourselves since hydrus server only stores the SHA-256 iirc. Could increase the sync time a lot.

That's sad.

**Comment by mserajnik** [comment]
> > But in that case we'd need to generate the MD5 ourselves since hydrus server only stores the SHA-256 iirc. Could increase the sync time a lot.
> 
> That's sad.

Well, I guess it's not necessary for it to store MD5 as well since the hydrus server database isn't usually queried directly. 😂

But I'll take a look tomorrow, I might be mistaken and there might be other hashes stored in a separate table.

**Comment by mserajnik** [comment]
> > Frontend (hydrusrvue)
> 
> I'd prefer using existing tag input with dropdown:
> 
> User starts typing **`"sys"`** -> show dropdown with **`"system"`** namespace
> User starts typing constraint type **`"system:n"`** -> show dropdown with **`"system:number of tags"`** constraint.
> User done typing constraint -> show dropdown with possible operators
> User done typing operator -> show dropdown with possible constraint arguments (in case of mime).
> User done typing constraint argument -> finish input and send the query.
> 
> Alot of benefits this way: you can simply copy-paste constraints, no special popups.
> 
> Also, i'd want to see added constraints as regular tags (like excluded tag, but with this [cute icon](https://fontawesome.com/icons/function?style=solid))
> 
> To help users, add help page with all possible queries (see [danbooru cheatsheet](https://danbooru.donmai.us/wiki_pages/43049))

This would need some special handling (so hitting enter on the dropdown doesn't add the tag if a constraint part is highlighted) but it's certainly doable and I like the idea. And if the user had tags that contain `system`, they'd be shown after the constraints in the dropdown ( so it's seamless and the user doesn't need to switch into a special mode).

The icon is good and distinguishable as well. Maybe call it _system functions_ instead of _system constraints_ then in terms of "official" terminology? _Constraints_ has a negative connotation anyway.

And yeah, I agree on the help page, with this and #17 this is getting too complicated for new users. I might even try to get that in before working on these two enhancements (maybe before the holidays).

**Comment by mserajnik** [comment]
> Still do not want use custom svg font? :D

No, I stand by the point I made in the other issue. :P

> I scraped all of these icons including pro.

You can add it in your fork if you want to (I don't condone it, but I can't prohibit it anyway), but I don't want anything that's technically illegal in my repos. Aside from the fact that I like supporting developers that are financially dependent on their service/product (even if I might not agree with Font Awesome's subscription model), two simple reasons:

- It would be a liability not only for myself but also the people hosting and using it themselves
- [Article 13 is scary](https://blog.github.com/2018-03-14-eu-proposal-upload-filters-code/)

**Comment by mserajnik** [comment]
Something like [`sliders-h`](https://fontawesome.com/icons/sliders-h?style=solid), [`tools`](https://fontawesome.com/icons/tools?style=solid) or maybe even [`question`](https://fontawesome.com/icons/question?style=solid) should be a decent enough alternative, even if I have to admit that `function` looks awesome.

**Comment by FredericaBernkastel** [comment]
[code](https://fontawesome.com/icons/code?style=solid)

**Comment by FredericaBernkastel** [comment]
I hope excluded constraints will be supported aswell.
Or atleast "not equals" operator.

**Comment by mserajnik** [comment]
> I hope excluded constraints will be supported aswell.
> Or atleast "not equals" operator.

Yes – `!=` will be supported, forgot to mention it, thanks.

**Comment by mserajnik** [comment]
Got a bit of a headstart on this. __No hydrusrvue support yet__, probably between Christmas and New Years' Eve (I hope).

First prototype: https://github.com/mserajnik/hydrusrv/commit/06035a3721b606e132d6c58acb41f1e773140c32

Didn't do extensive testing yet, so there might be some issues. Performance is pretty decent imo, especially considering the fact that most of the fields don't have indexes (and I'm not sure if it would be a good idea to index them all).

This needs [this hydrusrv-sync branch](https://github.com/mserajnik/hydrusrv-sync/tree/feature/constraints) to work.

Let me know what you think and if you find any issues.

**Comment by mserajnik** [comment]
Just realized: I didn't think about grouping together the constraints for each field with `OR`.

E.g., right now, `constraints[]=id=1000&constraints[]=id=2000` will return an empty set since no file can have two IDs. Instead, I should compare `id = 1000 || id = 2000` to make this far more useful.

Same with all the other fields, e.g., `constraints[]=width>5000&constraints[]=width<200` should filter for files that have either a `width > 5000` __or__ a `width < 200`.

Of course, providing two or more fields for constraints (e.g., for `width` and `height`) will still `AND` them, like with tags.

I don't think this will complicate things too much, I'll try to get this done tomorrow.

**Comment by FredericaBernkastel** [comment]
A small remark. Since you're using a smaller deviance (0.5 except 0.618), you should probably hardcode the result for 3 and 4.

https://github.com/mserajnik/hydrusrv/blob/ea50c78495c47b38a203732ad1de1379a03c7465/server/util/constraints.js#L87

**Comment by FredericaBernkastel** [comment]
I noticed there alot of functions for each combination of query types.
What about having a single framework to build such queries dynamically. It might look more complicated, but _technically_ - it's a better solution, and would be easier to implement new features in the future.

**Comment by FredericaBernkastel** [comment]
`> I chose 0.5 since 0.618 was a bit too much deviance for my taste at a value of 100, for example`
Or just let user specify the deviance by himself as an optional argument. E.g.
```
system:size~=1M,0.618
```

**Comment by mserajnik** [comment]
> > I chose 0.5 since 0.618 was a bit too much deviance for my taste at a value of 100, for example
> 
> Or just let the user to specify deviance by himself as an optional argument. E.g.
> 
> ```
> system:size~=1M,0.618
> ```

No, not intuitive to use, I'd rather have a fixed value that looks reasonable.

_But_, I've found another issue with the current system and coincidentally, I think the solution to this also solves this deviance problem:

Now, providing multiple constraints for a single field `OR`s them. Which makes sense for fields like `id` or `hash` (and all the other fields as well, when `=` or `!=` is used).

But for other fields like `width`, `OR` doesn't make sense when two constraints with opposing `>`/`<` operators are provided. In such a case, ideally that pairing should be handled with `AND`.

The tricky part is that the user can still provide multiple constraints with `>` and `<`, so it isn't clear anymore how to match them up. Going by the order provided doesn't sound like a great idea, because that would be a special rule just for this case as order is irrelevant everyhwere else in the API – unintuitive. And I don't want to constrict the user to one `>` and `<` constraint per field to work around this either.

Instead, I think the best solution is to add an additional comparator which combines `>` and `<`:

`width>100<200`

This way, the user can still provide `>` and `<` constraints as they are now (maybe he really wants to search files that are either `>1000` __or__ `<200`, who knows?) _and_ he can search in a certain range.

This is also a nice way to work around the `~=` deviance issue, as the user can now just provide a fixed range he deems appropriate.

**Comment by mserajnik** [comment]
Done, the `><` comparator is added: https://github.com/mserajnik/hydrusrv/commit/d1a0fadf0aaeab608938a8faedcbecb96c08b0f4

I think that should be it for this feature on the hydrusrv side. What's left are the necessary hydrusrvue changes/additions.

I thought about the whole query builder thing again and decided I'll put that off until this and #17 are released. The reason being that I don't yet exactly know how the queries in #17 will need to look like. Plus, if I also want to change the rest of the queries (not just files and tags), it's likely a pretty big task.

So, the final order will probably be:

- Release system constraints
- Release wildcard tags
- Finally work on #6
- Query builder + smaller refactoring/polish releases

**Comment by mserajnik** [comment]
I've made some more changes:

+ It's now possible to provide values for comparing with `size` as integer (for plain _bytes_) or integer/float + suffix `k`, `m`, `g` (for _kilobytes_, _megabytes_, _gigabytes_ – using `1000` for the multiplication, not `1024`, since more people will probably convert it that way instead of using kibibytes etc.), e.g., `width>5m`
+ When prodiving multiple `!=` constraints for a field, those are now compared with `AND` (instead of `OR`, like the other constraints for a field are), so it's possible to exclude multiple values while still including/comparing against multiple other values on the same field

I've also started working on the necessary hydrusrvue changes. No autocomplete for constraints yet, but setting constraints already works.

**Comment by FredericaBernkastel** [comment]
Never saw if someone measure filesize in base 1000. Even danbooru, claiming about `filesize:1M..1.001M` query is in megabytes, but still made in mebibytes.

You're not going to make a compromiss, like different suffix case, or exponential notation, right?

**Comment by mserajnik** [comment]
No, I don’t think having two solutions adds much, but confuses further.

But I‘m not biased, I mainly chose base 10 because many common softwares do it (e.g., `partman`) and it seemed easier for most people when calculating in their head.

Any particular reason why base 2 would be better?

Actually, I think we should probably use the same has hydrus in this case, so I‘ll check what‘s used there later.

**Comment by FredericaBernkastel** [comment]
In this case, why in the file page, you specify file size in base 2 instead 10?

**Comment by mserajnik** [comment]
> In this case, why in the file page, you specify file size in base 2 instead 10?

True, I would have needed to change it to base 10 there as well.

> Yes, it uses base 2 aswell.

Base 2 it is then. :)

Thanks for the feedback!

**Comment by mserajnik** [comment]
I should actually adjust [this](https://github.com/mserajnik/hydrusrvue/blob/2.2.1/src/components/file/Info.vue#L66) to use base 2 prefixes (`KiB`, `MiB` etc.) so it's correct.

Otherwise, we have the same situation as Windows where it calculates based on base 2 but displays base 10 prefixes (so, when Windows says `5 MB` in the explorer, it's actually `5 * 1024 * 1024` bytes, not `5 * 1000 * 1000`, so it should really say `5 MiB`).

**Comment by mserajnik** [comment]
hydrusrvue done as well: https://github.com/mserajnik/hydrusrvue/commit/02ca09b106353f06449f479b5dd06e42a6e75344

Aside from this main change, hydrusrvue now also allows for sorting by amount of tags (since hydrusrv supports this as well with the addition of the `tag_count` field). I've also added a new environment variable (`VUE_APP_SUBTITLE`) to (optionally) display a subtitle on the home view (for some added personal branding) and improved the autocomplete handling so it cancels unnecessary pending requests and no longer displays suggestions from requests that resolve after a search has been started (which was a potential issue before on slower connections).

Imo, this works pretty well with your suggestion of "partially completing" the constraints.

Easiest way to test all of this is by using the latest commit in the following branches:

- [hydrusrv/feature/constraints](https://github.com/mserajnik/hydrusrv/tree/feature/constraints)
- [hydrusrv-sync/feature/constraints](https://github.com/mserajnik/hydrusrv-sync/tree/feature/constraints)
- [hydrusrvue/feature/constraints](https://github.com/mserajnik/hydrusrvue/tree/feature/constraints)

Let me know if you find any issues.

The last thing to do in terms of "features" is to add the help page in hydrusrvue (I'll try to get this done today or tomorrow).

If everything goes well, I think I'll release this before the end of the year. :)

**Comment by mserajnik** [comment]
Small oversight: I forgot to allow comparisons against `0` which makes it difficult to filter for files without tags (only possible via `tags<1`) and impossible to filter for files that have at least one tag.

I'll correct this before release.

**Comment by FredericaBernkastel** [comment]
**1.** In the dropdown, icon is not vertically aligned with text.
![capture-129](https://user-images.githubusercontent.com/16236235/50424946-ae5e2480-0897-11e9-9a45-69352beccfe1.png)

Adding `display: flex` will fix one.
![capture-130](https://user-images.githubusercontent.com/16236235/50424948-bd44d700-0897-11e9-9446-cf9763dc77e5.png)

**2.** `VUE_APP_TITLE` and `VUE_APP_DESCRIPTION` are being lowercased while displayed on the homepage. While `<title>` and `<meta name=description>` still have the original case.
In the file page, size suffix is being lowercased aswell. `kib`, while `KiB` is more common.

**3.** In case of `mime` constraint, after typing the operator, it doesn't show the dropdown with available mime types. Did you forget about?

I'll test more a bit later.

**Comment by FredericaBernkastel** [comment]
**4.** **`size><2m,1m`** won't return any results. It should be treated as **`size>2m OR size<1m`**

**Comment by mserajnik** [comment]
1. You are right, it's an issue in Bulma. Actually, it's the same in the mobile navigation and the fix also works there. Seems like they forgot about icons in their dropdowns. Will need to add an additional `margin-right` to the icon as well (since wide icons will otherwise touch the text). I'll fix it.

<img width="119" alt="screenshot 2018-12-25 at 18 29 08" src="https://user-images.githubusercontent.com/28833155/50425262-0d5d7280-0873-11e9-96f3-f69bf0bd3b1c.png">

<img width="122" alt="screenshot 2018-12-25 at 18 29 16" src="https://user-images.githubusercontent.com/28833155/50425260-0b93af00-0873-11e9-8f5f-c02c0d3645c1.png">

2. Right, I'll lowercase `<title>` and `<meta name=description>` as well (since everything should be lowercase by design, as in hydrus itself) and make it so the size suffixes are in their regular case.

3. True, makes sense to add the ability to complete the mime types as well since it's just a fixed length list. I thought about adding a new route that returns the configured mime types __or__ make it necessary to configure them in hydrusrvue as well (so it will only list available ones and not all that hydrusrv supports in theory), but I'm not sure if either solution is worth it. Thoughts?

4. (you probably deleted the comment?) I think handling `size><2m,1m` the same as `size><1m,2m` (also for `><` comparisons on the other fields) would be best so the user doesn't have to think about the right order.

**Comment by FredericaBernkastel** [comment]
4. Yes, that's right.

**Comment by mserajnik** [comment]
Regarding 2.) once more: In hydrus client, you can configure it to be normal case instead of lowercase (lowercase being the default). I could add that option as well. Not sure if necessary/worth it?

**Comment by FredericaBernkastel** [comment]
`> Regarding 2.) once more: In hydrus client, you can configure it to be normal case instead of lowercase (lowercase being the default). I could add that option as well. Not sure if necessary/worth it?`

But what's the point? Just keep it in original case. User can lowercase it by himself.

**Comment by mserajnik** [comment]
No, I want it to be lowecase by default. If it‘s non-configurable, it‘s going to stay lowercase as it is.

**Comment by FredericaBernkastel** [comment]
**5.**
https://github.com/mserajnik/hydrusrvue/blob/2.2.1/src/views/File.vue#L48
https://github.com/mserajnik/hydrusrvue/blob/2.2.1/src/components/file/Info.vue#L30

That's not typically true, e.g. `video/x-msvideo`, `video/x-matroska`...
You should use a fixed extensions list, like i did.
FredericaBernkastel/hydrusrvue@2ee9cb3fffccbd40d0c7b7a8f6bbfb08128e3d9b

**Comment by mserajnik** [comment]
5. Weird that this never came up before. I'll change it so it adds a correct extension when downloading the file.

**Comment by mserajnik** [comment]
Closed by https://github.com/mserajnik/hydrusrv/commit/a69fa4e445a4c2b57c7bed6af47702eba958a1f9.
